### PR TITLE
Support for Manifest annotations in `bundle push`

### LINKED ISSUE #1933: Add support for annotating image manifest for OCI bundles
### Feature request

There are several annotations that provide useful metadata in the [OCI specification](https://github.com/opencontainers/image-spec/blob/main/annotations.md) that could be supported out of the box with sensible defaults. In addition to that I would like to provide my own annotations to be included when I push a OCI bundle via `tkn bundle push`.

### Use case

Additional metadata in the image would help me compare two images when the image is copied between registries and information contained within the image reference, e.g. name/version could be lost. This helps me maintain the metadata of the OCI bundle and helps me reason about provenance of a OCI bundle.

### UI Example

```
$ tkn bundle push \
    -a org.opencontainers.image.licenses=Apache-2.0 \
    --annotate 'org.opencontainers.image.title=My super bundle'
```

### PR DESCRIPTION
**Pull request by zregvart**
# Changes

Adds `--annotate` parameter to `tkn bundle push` to set any OCI Manifest annotations on the bundle image.

Fixes #1933

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Run the code checkers with `make check`
- [x] Regenerate the manpages, docs and go formatting with `make generated`
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

# Release Notes

```release-note
Add support for OCI Manifest Annotations via "--annotate" parameter to "tkn bundle push"
```

### CODE REVIEW

File: pkg/cmd/bundle/push.go [hunk 1]
```diff
@@ -80,6 +83,7 @@ Input:
 		},
 	}
 	c.Flags().StringSliceVarP(&opts.bundleContentPaths, "filenames", "f", []string{}, "List of fully-qualified file paths containing YAML or JSON defined Tekton objects to include in this bundle")
+	c.Flags().StringSliceVarP(&opts.annotationParams, "annotate", "a", []string{}, "OCI Manifest annotations in the form of key=value to be added to the OCI image")
```
Comment (piyush-garg) [inline pkg/cmd/bundle/push.go:86]: can you please add description like it covers multiple annotation scenario also

**Review by vdemeester — APPROVED** [review]

File: pkg/cmd/bundle/push.go (hunk 1 above)
Comment (vinamra28) [inline pkg/cmd/bundle/push.go:86]: `-a` flag sounds bit odd 😅 , let's keep `--annotate` only and not add any short-hand of this flag.
@piyush-garg @chmouel @vdemeester please let me know your thoughts on it

File: pkg/cmd/bundle/push.go [hunk 2]
```diff
@@ -109,6 +113,15 @@ func (p *pushOptions) parseArgsAndFlags(args []string) error {
 		p.bundleContents = append(p.bundleContents, string(contents))
 	}
 
+	p.annotations = map[string]string{}
+	for _, annParam := range p.annotationParams {
+		if k, v, ok := strings.Cut(annParam, "="); ok {
+			p.annotations[strings.TrimSpace(k)] = strings.TrimSpace(v)
+		} else {
```
Comment (vinamra28) [inline pkg/cmd/bundle/push.go:120]: can we make use of [ParseParams](https://github.com/tektoncd/cli/blob/main/pkg/params/mergeparams.go#L117) function here?

File: pkg/bundle/builder_test.go [hunk 3]
```diff
@@ -34,7 +34,8 @@ func TestBuildTektonBundle(t *testing.T) {
 		return
 	}
 
-	img, err := BuildTektonBundle([]string{string(raw)}, &bytes.Buffer{})
+	annotations := map[string]string{"a1": "v1", "a2": "v2"}
```
Comment (vinamra28) [inline pkg/bundle/builder_test.go:37]: let's use some meaningful annotations 😅

File: pkg/cmd/bundle/push_test.go [hunk 4]
```diff
@@ -71,6 +73,18 @@ func TestPushCommand(t *testing.T) {
 			stdin:            exampleTask,
 			expectedContents: map[string]expected{exampleTaskExpected.name: exampleTaskExpected},
 		},
+		{
+			name: "with-annotations",
+			files: map[string]string{
+				"simple.yaml": exampleTask,
+			},
+			annotations:      []string{"a1=k1", "a2 = k2"},
```
Comment (vinamra28) [inline pkg/cmd/bundle/push_test.go:81]: same, let's use some meaningful annotations

File: pkg/cmd/bundle/push.go (hunk 2 above)
Comment (zregvart) [reply pkg/cmd/bundle/push.go]: Refined `ParseParams` as well

File: pkg/params/mergeparams.go [hunk 5]
```diff
@@ -121,7 +121,12 @@ func ParseParams(params []string) (map[string]string, error) {
 		if len(r) != 2 {
 			return nil, errors.New(invalidParam + p)
 		}
-		parsedParams[r[0]] = r[1]
+		key := strings.TrimSpace(r[0])
+		val := strings.TrimSpace(r[1])
+		if key == "" {
+			return nil, errors.New(invalidParam + p)
+		}
```
Comment (vinamra28) [inline pkg/params/mergeparams.go:128]: @piyush-garg I don't recall why we didn't add this check at the first place, are having a case where key can be `""` 🤔

Comment (zregvart) [reply pkg/params/mergeparams.go:128]: I can't think of any usefulness of an empty key, JSON does allow for it. I thought being a bit more strict here is better at catching typos...

Comment (piyush-garg) [reply pkg/params/mergeparams.go:128]: I think it may be missed, but I don't think empty key can be of any use.

**Comment by piyush-garg** [comment]
lgtm
cc @vinamra28

**Comment by piyush-garg** [comment]
Please update commit message and PR desc for `-a` flag as it is removed now

**Comment by vinamra28** [comment]
> Please update commit message and PR desc for `-a` flag as it is removed now

@zregvart can you please take a look at this and then we can merge this PR

**Comment by zregvart** [comment]
Thanks!

**Pull request merged**